core/network/fabric-interop-cc/contracts/interop/manage_assets.go

```diff
@@ -338,3 +379,349 @@ func (s *SmartContract) ClaimAsset(ctx contractapi.TransactionContextInterface,
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	err = ctx.GetStub().DelState(assetLockKey)
+	if err != nil {
+		errorMessage := fmt.Sprintf("failed to delete lock for the asset associated with the contractId %s: %v", contractId, err)
+		log.Error(errorMessage)
+		return errors.New(errorMessage)
+	}
+
+	err = ctx.GetStub().DelState(contractIdPrefix + contractId)
+	if err != nil {
+		errorMessage := fmt.Sprintf("failed to delete the contractId %s as part of asset unlock: %v", contractId, err)
+		log.Error(errorMessage)
+		return errors.New(errorMessage)
+	}
+
+	return nil
+}
+
+// ClaimAsset cc is used to record claim of an asset on the ledger (this uses the contractId)
+func (s *SmartContract) ClaimAssetUsingContractId(ctx contractapi.TransactionContextInterface, contractId string, claimInfoBytes string) error {
+
+	assetLockKey, assetLockVal, err := fetchAssetLockedUsingContractId(ctx, contractId)
+	if err != nil {
+		log.Error(err.Error())
+		return err
+	}
+
+	claimInfo := &common.AssetClaimHTLC{}
+	err = proto.Unmarshal([]byte(claimInfoBytes), claimInfo)
+	if err != nil {
+		errorMsg := fmt.Sprintf("unmarshal error: %s", err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// display the claim information
+	log.Info(fmt.Sprintf("claimInfo: %+v\n", claimInfo))
+
+	// Check if expiry time is elapsed
+	currentTimeSecs := uint64(time.Now().Unix())
+	if uint64(currentTimeSecs) >= assetLockVal.ExpiryTimeSecs {
+		errorMsg := fmt.Sprintf("cannot claim asset associated with contractId %s as the expiry time is already elapsed", contractId)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	// compute the hash from the preimage
+	isCorrectPreimage, err := checkIfCorrectPreimage(string(claimInfo.HashPreimage), string(assetLockVal.Hash))
+	if err != nil {
+		errorMsg := fmt.Sprintf("claim asset associated with contractId %s failed with error: %v", contractId, err)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	if isCorrectPreimage == false {
+		errorMsg := fmt.Sprintf("cannot claim asset associated with contractId %s as the hash preimage is not matching", contractId)
+		log.Error(errorMsg)
+		return errors.New(errorMsg)
+	}
+
+	err = ctx.GetStub().DelState(assetLockKey)
+	if err != nil {
+		errorMessage := fmt.Sprintf("failed to delete lock for the asset associated with the contractId %s: %+v", contractId, err)
+		log.Error(errorMessage)
+		return errors.New(errorMessage)
+	}
+
+	err = ctx.GetStub().DelState(contractIdPrefix + contractId)
+	if err != nil {
+		errorMessage := fmt.Sprintf("failed to delete the contractId %s as part of asset claim: %+v", contractId, err)
+		log.Error(errorMessage)
+		return errors.New(errorMessage)
+	}
+
+	return nil
+}
+
+// IsAssetLocked cc is used to query the ledger and find out if an asset is locked or not (this uses the contractId)
+func (s *SmartContract) IsAssetLockedUsingContractId(ctx contractapi.TransactionContextInterface, contractId string) (bool, error) {
```

Comment:
The function name reads weird, as if we are asking whether the asset is locked using a contract Id. But I'm not sure what a suitable name would be.

Reply:
Changed the function name from `IsAssetLockedUsingContractId` to `IsAssetLockedQueryUsingContractId`.